sdks/java/extensions/timeseries/src/main/java/org/apache/beam/sdk/extensions/timeseries/joins/BiTemporalStreams.java

```diff
@@ -0,0 +1,464 @@
+          K key, List<TimestampedValue<V2>> right, TimestampedValue<V1> left, int idx) {
+
+        BiTemporalJoinResult<K, V1, V2> m = new BiTemporalJoinResult<>();
+        m.setLeftData(KV.of(key, left.getValue()), left.getTimestamp());
+
+        int adjustedIdx;
+        TimestampedValue<V2> value;
+
+        // If exact match found or if the largest timestamp in the list is smaller than the right
+        // streams timestamp id
+        // Note if there are duplicate for the same timestamp the results will be non-deterministic
+        if (idx >= 0) {
+          adjustedIdx = idx;
+          value = right.get(adjustedIdx);
+          m.setRightData(KV.of(key, value.getValue()), value.getTimestamp());
+          m.setMatched(true);
+        }
+        if (right.size() > 0 && idx == -1 - right.size()) {
+          adjustedIdx = right.size() - 1;
+          value = right.get(adjustedIdx);
+          m.setRightData(KV.of(key, value.getValue()), value.getTimestamp());
+          m.setMatched(true);
+        }
+        if (idx < -1) {
+          adjustedIdx = Math.abs(idx) - 2;
+          value = right.get(adjustedIdx);
+          m.setRightData(KV.of(key, value.getValue()), value.getTimestamp());
+          m.setMatched(true);
+        }
+        if (idx == -1) {
+          m.setMatched(false);
+        }
+
+        return m;
+      }
+
+      /*
+       Right stream values are added to the bag
+       Left stream values are added to the bag
+       A timer is set if no timer exists already or if min(timerTimestamp) is > current object timestamp
+
+       If timertimestamp is NULL then a timer is set in OnProcess()
+       If On OnTimer firing there are elements in the left stream where  EventTime > OnTimer.timestamp then a new timer is also set.
+      */
+      @ProcessElement
+      public void process(
+          @Element KV<K, BiTemporalJoinResult<K, V1, V2>> input,
+          @Timestamp Instant timestamp,
+          @StateId("rightStream") BagState<TimestampedValue<V2>> rightStream,
+          @StateId("leftStream") BagState<TimestampedValue<V1>> leftStream,
+          @StateId("timerTimestamp") ValueState<Long> timerTimestamp,
+          @StateId("dataKey") ValueState<K> dataKey,
+          @TimerId("processTimer") Timer processTimer) {
+
+        // This workaround is due to OnTimerContext not supporting key access
+        if (dataKey.read() == null) {
+          dataKey.write(input.getKey());
+        }
+
+        if (input.getValue().getRightData() != null) {
+          TimestampedValue<V2> rightData = input.getValue().getRightData();
+          rightStream.add(rightData);
+        }
+
+        if (input.getValue().getLeftData() != null) {
+          TimestampedValue<V1> rightStreamValue = input.getValue().getLeftData();
+          leftStream.add(rightStreamValue);
+
+          long processTimerTimestamp = Optional.ofNullable(timerTimestamp.read()).orElse(0L);
```

Comment:
Timestamp 0 seems no to be ideal default value, because it can (in theory) be valid timestamp that was actually set by an element. Maybe `BoundedWindow.TIMESTAMP_MIN_VALUE.getMillis()` would be better. And maybe instead of `long` the state could hold the `Instant`? `InstantCoder` should be efficient enough.